File: HW3/decipher.py
def decipher(secret, N, d):
    result = []
    for chunk in secret:
        chunk = pow(chunk, d, N)
        result.append(chr(chunk >> 8))
        result.append(chr(chunk & 0b11111111))
    if result[-1] == chr(0):
        result.pop()
    return result

if __name__ == '__main__':
    with open('secret.txt', 'r') as f:
        secret = map(int, f.read().strip().split('\n'))
    with open('private_key.txt', 'r') as f:
        N = int(f.readline().strip())
        d = int(f.readline().strip())
    with open('message.txt', 'w') as f:
        for char in decipher(secret, N, d):
            f.write(char)

File: HW3/decipher.cpp
#include <cstdint>
#include <fstream>
#include "pow.h"

template<class T, class = typename std::enable_if<std::is_integral<T>::value>::type>
void decrypt(std::ofstream& ofs, std::ifstream& ifs, T N, T d) {
    T chunk = 0;
    char temp = 0;
    while(ifs >> chunk) {
        chunk = pow(chunk, d, N);
        ofs << static_cast<char>(chunk >> 8);
        temp = chunk & 0xFF;
        if(ifs.peek() != EOF || temp) // is not last or (is last chunk and temp is not 0)
            ofs << temp;
    }
    ofs.flush();
}

int main() {
    std::ifstream msg("secret.txt"), pub("private_key.txt");
    std::ofstream ofs("message.txt");
    uint64_t N, d;
    pub >> N >> d;
    decrypt(ofs, msg, N, d);
    return 0;
}